### ISSUE TITLE
Postponed extended lambda leads to ambiguous extendedlambda-name

### ISSUE THREAD

**Issue by mietek** [agda/agda#2028]
> An internal error has occurred. Please report this as a bug.
> Location of the error: `src/full/Agda/TypeChecking/Reduce/Monad.hs:176`

This happens when attempting to fill either of the holes (`C`-`space`).

``` agda
open import Data.Empty using (⊥ ; ⊥-elim)
open import Data.Product using (_×_ ; _,_)
open import Data.Sum using (_⊎_ ; inj₁ ; inj₂)
open import Data.Unit using (⊤ ; tt)
open import Relation.Nullary using (¬_)

Subset : Set → Set₁
Subset X = X → Set

_∈_ : ∀ {X} → X → Subset X → Set
x ∈ A = A x

_∉_ : ∀ {X} → X → Subset X → Set
x ∉ A = ¬ x ∈ A

∅ : ∀ {X} → Subset X
∅ = {!λ _ → ⊥!}

∅-empty : ∀ {X} {x : X} → x ∉ ∅
∅-empty = ⊥-elim

𝕌 : ∀ {X} → Subset X
𝕌 = {!λ _ → ⊤!}

𝕌-univ : ∀ {X} {x : X} → x ∈ 𝕌
𝕌-univ = tt

_⊆_ : ∀ {X} → Subset X → Subset X → Set
A ⊆ B = ∀ {x} → x ∈ A → x ∈ B

_⫗_ : ∀ {X} → Subset X → Subset X → Set
A ⫗ B = A ⊆ B × B ⊆ A

_∪_ : ∀ {X} → Subset X → Subset X → Subset X
A ∪ B = λ x → x ∈ A ⊎ x ∈ B

∪-bound : ∀ {X} {A : Subset X} → (A ∪ 𝕌) ⫗ 𝕌
∪-bound = (λ { (inj₁ x∈A) → tt
             ; (inj₂ tt) → tt
             }) ,
          (λ tt → inj₂ tt)
```

**Comment by UlfNorell** [comment]
Here's a simplified and self-contained test case:

``` agda
module _ where

record ⊤ : Set where
  constructor tt

data Either (A B : Set) : Set where
  inl : A → Either A B
  inr : B → Either A B

Subset : Set → Set₁
Subset X = X → Set

U : ∀ {X} → Subset X
U _ = {!⊤!}

_⊆_ : ∀ {X} → Subset X → Subset X → Set
A ⊆ B = ∀ {x} → A x → B x

_∪_ : ∀ {X} → Subset X → Subset X → Subset X
(A ∪ B) x = Either (A x) (B x)

∪-bound : ∀ {X} {A : Subset X} → (A ∪ U) ⊆ U
∪-bound =
  λ { (inl _)  → tt
    ; (inr tt) → tt
    }
```

It looks like the problem is that the pattern lambda is getting postponed, but I have no idea how that is causing the internal error. The `__IMPOSSIBLE__` is triggered because the pattern lambda function is both in the signature for the current file and in the imported signature (where it really shouldn't be).

**Comment by andreasabel** [comment]
With `{-# OPTIONS -v import:100 #-}` we see

```
Finished type checking.
Starting serialization.
Building interface...
  instantiating all meta variables
  interface complete
Finished serialization.
We have unsolved metas.
?0 : Set
_31 : (A ∪ U) .x → U .x  [ at /home/abel/agda/test/interaction/Issue2028.agda:24,3-26,6 ]
Starting writing to interface file.
Finished writing to interface file.
  New module. Let's check it out.
  Now we've looked at Issue2028
```

This means:
- `mergeInterface` is not called, so `stateChangesIncluded` must be `True` (as `visited` is `False`)
- but `typeCheckThe file` calls `addImportedThings` which merges the `stSignature` into the `stImports`
  So maybe it is not surprising that the name appears both in the signature and the imported signature.

**Comment by andreasabel** [comment]
#2093 is another instance of this bug.

**Comment by andreasabel** [comment]
Ha, found the culprit, I think.  Mea culpa.  See `Term.hs`:

``` haskell
       -- Case: we could not check the extended lambda because we are blocked on a meta.
       -- In this case, we want to postpone.
       Just (err, x) -> do
         -- Note that we messed up the state a bit.  We might want to unroll these state changes.
         -- However, they are harmless:
         -- 1. We created a new mutual block id.
         -- 2. We added a constant without definition.
         -- TODO: roll back the state.
```

Typical final words...

**Comment by andreasabel** [comment]
We fix this by resetting the state to before the attempted type-checking of the extended lambda (which is being postponed).
